# 1. Splitting/ chunking
---------------------------------------------------------------------------------

<pre>
- Chunk_size: 
    . too slow: difficult to interpret
    . too big: lack context

- Chunk_overlap: include informations beyond the boundry

### a.) ChracterTextSplitter

In [7]:
from langchain_text_splitters import CharacterTextSplitter

text = """Lorem Ipsum is simply dummy text of the printing and typesetting industry.\n\n
Lorem Ipsum has been the industry's standard dummy text ever since 1966, 
\n\nIt has survived not only many decades, but also the leap into electronic typesetting, 
\n\nIt was popularised thanks to these sheets and more 
recently with desktop publishing software like Aldus PageMaker and Microsoft Word \n\n
including versions of Lorem Ipsum.

 """

# initialiser le CharacterTextSplitter
text_splitter = CharacterTextSplitter(separator= "\n\n",
                                    chunk_size= 100,
                                    chunk_overlap= 10)



In [8]:
chunks = text_splitter.split_text(text)

Created a chunk of size 135, which is longer than the specified 100


In [9]:
chunks

['Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
 "Lorem Ipsum has been the industry's standard dummy text ever since 1966,",
 'It has survived not only many decades, but also the leap into electronic typesetting,',
 'It was popularised thanks to these sheets and more \nrecently with desktop publishing software like Aldus PageMaker and Microsoft Word',
 'including versions of Lorem Ipsum.']

In [11]:
print([len(chunk) for chunk in chunks])

[74, 72, 85, 133, 34]


- chunks may lack context
- chunks may be larger than chunk_size specified 133>>> 100 !! To solve this we can use :

<b> RecursiveCharacterTextSplitter</b>


### b.) RecursiveCharacterTextSplitter

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

splitter = RecursiveCharacterTextSplitter(
    separators= ["\n\n", "\n", " ", ""],  # provide a list of seperators
    chunk_size= 100,
    chunk_overlap = 10
)


# On fait le split selon le premier seperator: "\n\n" si la taille <= chunk_size c'est bon, sinon:
# On fait le split selon le deuxième seperator: "\n" si la taille <= chunk_size c'est bon, sinon:
# On fait le split selon le troisième seperator: " " si la taille <= chunk_size c'est bon, sinon:
# # On fait le split selon le dernier seperator: ""
chunks = splitter.split_text(text)
chunks

['Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
 "Lorem Ipsum has been the industry's standard dummy text ever since 1966,",
 'It has survived not only many decades, but also the leap into electronic typesetting,',
 'It was popularised thanks to these sheets and more',
 'recently with desktop publishing software like Aldus PageMaker and Microsoft Word',
 'including versions of Lorem Ipsum.']

In [14]:
print([len(chunk) for chunk in chunks]) # Now all the chunks will have a size <= chunk_size

[74, 72, 85, 50, 81, 34]


In [ ]:
#It was popularised thanks to these sheets and more \nrecently with desktop publishing software like Aldus PageMaker and Microsoft Word',
#'including versions of Lorem Ipsum.

### c.) Splitting documents

In [17]:
from langchain_community.document_loaders import PyPDFLoader

loader = PyPDFLoader("pdf_file_example.pdf")
documents = loader.load()

splitter = RecursiveCharacterTextSplitter(
    chunk_size = 1000,
    chunk_overlap= 200
)


chunks = splitter.split_documents(documents)

chunks

[Document(page_content='Lorem ipsum \nLorem ipsum dolor sit amet, consectetur adipiscing \nelit. Nunc ac faucibus odio. \nVestibulum neque massa, scelerisque sit amet ligula eu, congue molestie mi. Praesent ut\nvarius sem. Nullam at porttitor arcu, nec lacinia nisi. Ut ac dolor vitae odio interdum\ncondimentum.  Vivamus  dapibus  sodales  ex,  vitae  malesuada  ipsum  cursus\nconvallis. Maecenas sed egestas nulla, ac condimentum orci.  Mauris diam felis,\nvulputate ac suscipit et, iaculis non est. Curabitur semper arcu ac ligula semper, nec luctus\nnisl blandit. Integer lacinia ante ac libero lobortis imperdiet. Nullam mollis convallis ipsum,\nac accumsan nunc vehicula vitae. Nulla eget justo in felis tristique fringilla. Morbi sit amet\ntortor quis risus auctor condimentum. Morbi in ullamcorper elit. Nulla iaculis tellus sit amet\nmauris tempus fringilla.\nMaecenas mauris lectus, lobortis et purus mattis, blandit dictum tellus.\n\uf0b7 Maecenas non lorem quis tellus placerat varius. \

In [20]:
print([len(chunk.page_content) for chunk in chunks])

[975, 502, 992, 984, 331, 980, 973, 963, 603]


# 2. Embedding
---------------------------------------------------------------------------------

In [ ]:
# We use openAi model for embeddings
# and ChromaDB for storing
# Before using this install: pip install langchain-openai
# also install : pip install langchain-chroma


from langchain_openai import OpenAIEmbeddings
from langchain_chroma import Chroma

# Initialiser le modèle d'embedding "clé +  le nom du modèle"
embedding_model = OpenAIEmbeddings(
    api_key = " "  # we need to add the correct key for the paid model !!!,
    model= "text-embedding-3-small"
)

# 3. Storing
---------------------------------------------------------------------------------

In [23]:
vector_store = Chroma.from_documents(
    documents= chunks,
    embedding= embedding_model
)

# store the embeddings in chroma db

NameError: name 'Chroma' is not defined